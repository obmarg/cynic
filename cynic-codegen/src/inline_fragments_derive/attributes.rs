use crate::attributes::{extract_meta_attrs, Attribute};

#[derive(Debug)]
pub struct InlineFragmentsDeriveAttributes {
    pub schema_path: Attribute,
    pub query_module: Attribute,
    pub graphql_type: Attribute,
    pub argument_struct: Option<Attribute>,
}

pub fn parse(attrs: &Vec<syn::Attribute>) -> Result<InlineFragmentsDeriveAttributes, syn::Error> {
    let (mut attr_map, attr_span) = extract_meta_attrs::<DeriveAttribute>(attrs)?;

    let schema_path = attr_map
        .remove(&DeriveAttribute::SchemaPath)
        .ok_or(syn::Error::new(
            attr_span,
            "Missing required attribute: schema_path",
        ))?;

    let query_module = attr_map
        .remove(&DeriveAttribute::QueryModule)
        .ok_or(syn::Error::new(
            attr_span,
            "Missing required attribute: query_module",
        ))?;

    let graphql_type = attr_map
        .remove(&DeriveAttribute::GraphqlType)
        .ok_or(syn::Error::new(
            attr_span,
            "Missing required attribute: graphql_type",
        ))?;

    let argument_struct = attr_map.remove(&DeriveAttribute::ArgumentStruct);

    Ok(InlineFragmentsDeriveAttributes {
        schema_path,
        query_module,
        graphql_type,
        argument_struct,
    })
}

#[derive(Debug, PartialEq, Eq, Hash)]
enum DeriveAttribute {
    SchemaPath,
    QueryModule,
    GraphqlType,
    ArgumentStruct,
}

impl std::str::FromStr for DeriveAttribute {
    type Err = String;

    fn from_str(s: &str) -> Result<DeriveAttribute, String> {
        if s == "schema_path" {
            Ok(DeriveAttribute::SchemaPath)
        } else if s == "query_module" {
            Ok(DeriveAttribute::QueryModule)
        } else if s == "graphql_type" {
            Ok(DeriveAttribute::GraphqlType)
        } else if s == "argument_struct" {
            Ok(DeriveAttribute::ArgumentStruct)
        } else {
            Err(format!("Unknown cynic attribute: {}.  Expected one of schema_path, query_module, graphql_type or argument_struct", s))
        }
    }
}
