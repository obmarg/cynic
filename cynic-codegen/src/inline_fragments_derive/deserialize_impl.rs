use proc_macro2::TokenStream;

pub struct DeserializeImpl<'a> {
    pub(super) target_enum: syn::Ident,
    pub(super) fragments: &'a [super::Fragment],
    pub(super) fallback: Option<(syn::Ident, Option<syn::Type>)>,
}

impl quote::ToTokens for DeserializeImpl<'_> {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        use quote::{quote, TokenStreamExt};

        let target_enum = &self.target_enum;
        let inner_types = self.fragments.iter().map(|fragment| &fragment.inner_type);
        let graphql_types = self
            .fragments
            .iter()
            .map(|fragment| proc_macro2::Literal::string(&fragment.graphql_type));
        let variant_names = self
            .fragments
            .iter()
            .map(|fragment| &fragment.rust_variant_name)
            .collect::<Vec<_>>();

        let (intermediate_fallback_def, fallback_match) = match &self.fallback {
            Some((fallback_variant, Some(fallback_inner))) => (
                Some(quote! { #[serde(other)] Fallback(#fallback_inner)}),
                Some(
                    quote! { Intermediate::Fallback(inner) => #target_enum::#fallback_variant(inner) },
                ),
            ),
            Some((fallback_variant, None)) => (
                Some(quote! { #[serde(other)] Fallback }),
                Some(quote! { Intermediate::Fallback => #target_enum::#fallback_variant }),
            ),
            None => (None, None),
        };

        tokens.append_all(quote! {
            #[automatically_derived]
            impl<'de> ::cynic::serde::Deserialize<'de> for #target_enum {
                fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
                where
                    D: ::cynic::serde::Deserializer<'de>,
                {
                    #[derive(::cynic::serde::Deserialize)]
                    #[serde(tag = "__typename", crate="::cynic::serde")]
                    enum Intermediate {
                        #(
                            #[serde(rename = #graphql_types)]
                            #variant_names(#inner_types),
                        )*
                        #intermediate_fallback_def
                    }

                    Ok(match Intermediate::deserialize(deserializer)? {
                        #(
                            Intermediate::#variant_names(inner) => #target_enum::#variant_names(inner),
                        )*
                        #fallback_match
                    })
                }
            }
        });
    }
}
