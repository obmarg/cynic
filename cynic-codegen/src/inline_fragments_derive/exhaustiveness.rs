use std::collections::HashSet;

use darling::util::SpannedValue;
use proc_macro2::Span;

use crate::{
    error::Errors,
    suggestions::{format_guess, guess_field},
};

use super::{input::InlineFragmentsDeriveVariant, InlineFragmentType};

pub(super) fn exhaustiveness_check(
    variants: &[SpannedValue<InlineFragmentsDeriveVariant>],
    target_type: &InlineFragmentType<'_>,
) -> Result<(), Errors> {
    let variant_names = variants
        .iter()
        .filter(|v| !*v.fallback)
        .map(|v| v.ident.to_string())
        .collect::<HashSet<_>>();

    let InlineFragmentType::Union(union_type) = target_type else {
        // I can't be bothered implementing exhaustiveness for interfaces now
        // so I'm just making it unsupported. If anyone wants it they're welcome to
        // contribute an implementation
        unreachable!()
    };

    let required_variants = union_type
        .types
        .iter()
        .map(|ty| ty.to_string())
        .collect::<HashSet<_>>();

    if variant_names != required_variants {
        let mut errors = Errors::default();

        for unexpected_variant_name in variant_names.difference(&required_variants) {
            let variant = variants
                .iter()
                .find(|v| v.ident == *unexpected_variant_name)
                .unwrap();
            let guess_field = guess_field(
                required_variants
                    .iter()
                    .map(|variant_name| variant_name.as_str()),
                unexpected_variant_name,
            );
            errors.push(syn::Error::new(
                variant.span(),
                format!(
                    "Could not find a member named {} in the union {}. {}",
                    unexpected_variant_name,
                    target_type.name(),
                    format_guess(guess_field)
                ),
            ));
        }

        for missing_variant_name in required_variants.difference(&variant_names) {
            errors.push(syn::Error::new(
                Span::call_site(),
                format!(
                    "This InlineFragment is missing a variant for {}.  Either provide a variant for this type or remove the exhaustive attribute",
                    missing_variant_name
                ),
            ));
        }

        return Err(errors);
    }

    Ok(())
}
