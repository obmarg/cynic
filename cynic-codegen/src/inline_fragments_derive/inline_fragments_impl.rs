use proc_macro2::TokenStream;

pub struct InlineFragmentsImpl<'a> {
    pub(super) target_enum: syn::Ident,
    pub(super) fragments: &'a [super::Fragment],
    pub(super) fallback: Option<(syn::Ident, Option<syn::Type>)>,
}

impl quote::ToTokens for InlineFragmentsImpl<'_> {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        use quote::{quote, TokenStreamExt};

        let target_enum = &self.target_enum;
        let inner_types = self.fragments.iter().map(|fragment| &fragment.inner_type);
        let variant_names = self
            .fragments
            .iter()
            .map(|fragment| &fragment.rust_variant_name)
            .collect::<Vec<_>>();

        let fallback = match &self.fallback {
            Some((variant, None)) => quote! {
                Ok(#target_enum::#variant)
            },
            Some((variant, Some(ty))) => quote! {
                <#ty as ::cynic::serde::Deserialize<'de>>::deserialize(deserializer).map(
                    #target_enum::#variant
                )
            },
            None => {
                quote! {
                    use ::cynic::serde::de::Error;
                    Err(D::Error::custom(format!("Unknown type: {}", typename)))
                }
            }
        };

        tokens.append_all(quote! {
            #[automatically_derived]
            impl<'de> ::cynic::serde::Deserialize<'de> for #target_enum {
                fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
                where
                    D: ::cynic::serde::Deserializer<'de>,
                {
                    deserializer.deserialize_map(::cynic::__private::InlineFragmentVisitor::<Self>::new())
                }
            }

            #[automatically_derived]
            impl<'de> ::cynic::InlineFragments<'de> for #target_enum {
                fn deserialize_variant<D>(typename: &str, deserializer: D) -> Result<Self, D::Error>
                where
                    D: ::cynic::serde::Deserializer<'de>
                {
                    #(
                        if Some(typename) == <#inner_types as ::cynic::QueryFragment<'de>>::TYPE {
                            return <#inner_types as ::cynic::serde::Deserialize<'de>>::deserialize(deserializer).map(
                                #target_enum::#variant_names
                            )
                        }
                    )*

                    #fallback
                }
            }
        });
    }
}
