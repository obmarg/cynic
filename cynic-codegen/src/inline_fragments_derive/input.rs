use darling::util::SpannedValue;
use proc_macro2::Span;

#[derive(darling::FromDeriveInput)]
#[darling(attributes(cynic), supports(enum_newtype, enum_unit))]
pub struct InlineFragmentsDeriveInput {
    pub(super) ident: proc_macro2::Ident,
    pub(super) data: darling::ast::Data<SpannedValue<InlineFragmentsDeriveVariant>, ()>,

    pub schema_path: SpannedValue<String>,

    // query_module is deprecated, remove eventually.
    #[darling(default)]
    query_module: Option<SpannedValue<String>>,
    #[darling(default, rename = "schema_module")]
    schema_module_: Option<syn::Path>,

    #[darling(default)]
    pub graphql_type: Option<SpannedValue<String>>,
    #[darling(default)]
    pub argument_struct: Option<syn::Ident>,
}

impl InlineFragmentsDeriveInput {
    pub fn schema_module(&self) -> syn::Path {
        if let Some(schema_module) = &self.schema_module_ {
            return schema_module.clone();
        }
        if let Some(query_module) = &self.query_module {
            return syn::parse_str(query_module).unwrap();
        }
        syn::parse2(quote::quote! { schema }).unwrap()
    }

    pub fn graphql_type_name(&self) -> String {
        self.graphql_type
            .as_ref()
            .map(|sp| sp.to_string())
            .unwrap_or_else(|| self.ident.to_string())
    }

    pub fn graphql_type_span(&self) -> Span {
        self.graphql_type
            .as_ref()
            .map(|val| val.span())
            .unwrap_or_else(|| self.ident.span())
    }
}

#[derive(darling::FromVariant)]
#[darling(attributes(cynic))]
pub(super) struct InlineFragmentsDeriveVariant {
    pub(super) ident: proc_macro2::Ident,
    pub fields: darling::ast::Fields<InlineFragmentsDeriveField>,

    #[darling(default)]
    rename: Option<SpannedValue<String>>,

    #[darling(default)]
    pub(super) fallback: SpannedValue<bool>,
}

#[derive(darling::FromField)]
#[darling(attributes(cynic))]
pub(super) struct InlineFragmentsDeriveField {
    pub ty: syn::Type,
}

impl InlineFragmentsDeriveVariant {
    pub(super) fn graphql_ident(&self) -> crate::Ident {
        if let Some(rename) = &self.rename {
            return crate::Ident::for_type(&**rename).with_span(rename.span());
        }

        crate::Ident::from_proc_macro2(&self.ident, None)
    }
}
