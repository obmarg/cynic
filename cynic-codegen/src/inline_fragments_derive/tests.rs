use std::{io::Write, process::Stdio};

use insta::assert_snapshot;
use rstest::rstest;
use syn::parse_quote;

use super::inline_fragments_derive;

#[rstest(input => [
    // Test of a union type.
    parse_quote!(
        #[derive(InlineFragments, Serialize)]
        #[cynic(schema_path = "../cynic/tests/test-schema.graphql")]
        enum PostOrAuthor {
            #[cynic(rename = "BlogPost")]
            Post(Post),
            Author(Author),
        }
    ),
    // Union type with a fallback
    parse_quote!(
        #[derive(InlineFragments, Serialize)]
        #[cynic(schema_path = "../cynic/tests/test-schema.graphql")]
        enum PostOrAuthor {
            #[cynic(rename = "BlogPost")]
            Post(Post),
            Author(Author),
            #[cynic(fallback)]
            Other
        }
    ),
    // Interface
    parse_quote!(
        #[derive(InlineFragments, Serialize)]
        #[cynic(schema_path = "../cynic/tests/test-schema.graphql")]
        enum Node {
            #[cynic(rename = "BlogPost")]
            Post(Post),
            Author(Author),
            #[cynic(fallback)]
            Other
        }
    ),
])]
fn snapshot_inline_fragments_derive(input: syn::DeriveInput) {
    let tokens = inline_fragments_derive(&input).unwrap();

    assert_snapshot!(format_code(format!("{}", tokens)));
}

fn format_code(input: String) -> String {
    let mut cmd = std::process::Command::new("rustfmt")
        .stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .stderr(Stdio::inherit())
        .spawn()
        .expect("failed to execute rustfmt");

    write!(cmd.stdin.as_mut().unwrap(), "{}", input).unwrap();

    std::str::from_utf8(&cmd.wait_with_output().unwrap().stdout)
        .unwrap()
        .to_owned()
}
