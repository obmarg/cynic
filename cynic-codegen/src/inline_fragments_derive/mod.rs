use darling::util::SpannedValue;
use proc_macro2::{Span, TokenStream};

use crate::{idents::PathExt, load_schema, schema, Errors, Ident};

pub mod input;

mod inline_fragments_impl;

#[cfg(test)]
mod tests;

pub use input::InlineFragmentsDeriveInput;

use crate::suggestions::{format_guess, guess_field};
use input::InlineFragmentsDeriveVariant;
use std::collections::HashSet;

pub fn inline_fragments_derive(ast: &syn::DeriveInput) -> Result<TokenStream, Errors> {
    use darling::FromDeriveInput;

    match InlineFragmentsDeriveInput::from_derive_input(ast) {
        Ok(input) => inline_fragments_derive_impl(input),
        Err(e) => Ok(e.write_errors()),
    }
}

pub(crate) fn inline_fragments_derive_impl(
    input: InlineFragmentsDeriveInput,
) -> Result<TokenStream, Errors> {
    use quote::{quote, quote_spanned};

    let schema =
        load_schema(&*input.schema_path).map_err(|e| e.into_syn_error(input.schema_path.span()))?;

    let target_type = find_union_or_interface_type(&input.graphql_type_name(), &schema);
    if target_type.is_none() {
        use graphql_parser::schema::{Definition, TypeDefinition};
        let candidates = schema.definitions.iter().flat_map(|def| match def {
            Definition::TypeDefinition(TypeDefinition::Union(union)) => Some(union.name.as_str()),
            Definition::TypeDefinition(TypeDefinition::Interface(interface)) => {
                Some(interface.name.as_str())
            }
            _ => None,
        });
        let guess_field = guess_field(candidates, &(input.graphql_type_name()));
        return Err(syn::Error::new(
            input.graphql_type_span(),
            format!(
                "Could not find a Union type or Interface named {}.{}",
                &input.graphql_type_name(),
                format_guess(guess_field)
            ),
        )
        .into());
    }
    let target_type = target_type.unwrap();

    let input_argument_struct = (&input.argument_struct).clone();
    let argument_struct = if let Some(arg_struct) = input_argument_struct {
        let span = arg_struct.span();
        let arg_struct_val: Ident = arg_struct.into();
        let argument_struct = quote_spanned! { span => #arg_struct_val };
        syn::parse2(argument_struct)?
    } else {
        syn::parse2(quote! { () })?
    };

    if let darling::ast::Data::Enum(variants) = &input.data {
        exhaustiveness_check(variants, &target_type, &schema)?;

        let fallback = check_fallback(variants, &target_type)?;

        let mut type_lock = input.schema_module();
        type_lock.push(Ident::for_type(input.graphql_type_name()));

        let fragments = fragments_from_variants(variants)?;

        let query_fragment_impl = QueryFragmentImpl {
            target_enum: input.ident.clone(),
            type_lock,
            argument_struct,
            fragments: &fragments,
            graphql_type_name: input.graphql_type_name(),
            fallback: fallback.clone(),
        };

        let inline_fragments_impl = inline_fragments_impl::InlineFragmentsImpl {
            target_enum: input.ident.clone(),
            fragments: &fragments,
            fallback,
        };

        Ok(quote! {
            #inline_fragments_impl
            #query_fragment_impl
        })
    } else {
        Err(syn::Error::new(
            Span::call_site(),
            "InlineFragments can only be derived from an enum".to_string(),
        )
        .into())
    }
}

fn exhaustiveness_check(
    variants: &[SpannedValue<InlineFragmentsDeriveVariant>],
    target_type: &InlineFragmentType,
    schema: &schema::Document,
) -> Result<(), Errors> {
    use schema::{Definition, TypeDefinition};

    let variant_names = variants
        .iter()
        .filter(|v| !*v.fallback)
        .map(|v| Ident::for_type(v.graphql_name()))
        .collect::<HashSet<_>>();

    let required_variants = match target_type {
        InlineFragmentType::Interface(iface) => schema
            .definitions
            .iter()
            .map(|d| match d {
                Definition::TypeDefinition(TypeDefinition::Object(obj)) => {
                    if obj.implements_interfaces.contains(&iface.name) {
                        Some(Ident::for_type(&obj.name))
                    } else {
                        None
                    }
                }
                _ => None,
            })
            .flatten()
            .collect::<HashSet<_>>(),
        InlineFragmentType::Union(union) => union
            .types
            .iter()
            .map(Ident::for_type)
            .collect::<HashSet<_>>(),
    };

    let has_fallback = variants.iter().any(|v| *v.fallback);

    if has_fallback && !variant_names.is_subset(&required_variants) {
        let mut errors = Errors::default();

        for unexpected_variant_name in variant_names.difference(&required_variants) {
            let variant = variants
                .iter()
                .find(|v| Ident::for_type(v.graphql_name()) == *unexpected_variant_name)
                .unwrap();

            let candidates = required_variants.iter().map(|v| v.graphql_name());
            let guess_field = guess_field(candidates, &variant.graphql_name());
            errors.push(syn::Error::new(
                variant.span(),
                format!(
                    "Could not find a match for {} in {}.{}",
                    variant.graphql_name(),
                    target_type.name(),
                    format_guess(guess_field)
                ),
            ))
        }

        return Err(errors);
    } else if !has_fallback && variant_names != required_variants {
        let mut errors = Errors::default();

        for unexpected_variant_name in variant_names.difference(&required_variants) {
            let variant = variants
                .iter()
                .find(|v| Ident::for_type(v.graphql_name()) == *unexpected_variant_name)
                .unwrap();
            let candidates = required_variants.iter().map(|v| v.graphql_name());
            let guess_field = guess_field(candidates, &variant.graphql_name());
            errors.push(syn::Error::new(
                variant.span(),
                format!(
                    "Could not find a match for {} in {}.{}",
                    variant.graphql_name(),
                    target_type.name(),
                    format_guess(guess_field)
                ),
            ));
        }

        for missing_variant_name in required_variants.difference(&variant_names) {
            errors.push(syn::Error::new(
                Span::call_site(),
                format!(
                    "This InlineFragment is missing a variant for {}.  Either provide a variant for this type or add a fallback variant.",
                    missing_variant_name.graphql_name()
                ),
            ));
        }

        return Err(errors);
    }

    Ok(())
}

struct Fragment {
    rust_variant_name: syn::Ident,
    inner_type: syn::Type,
}

fn fragments_from_variants(
    variants: &[SpannedValue<InlineFragmentsDeriveVariant>],
) -> Result<Vec<Fragment>, syn::Error> {
    let mut result = vec![];
    for variant in variants {
        if *variant.fallback {
            continue;
        }

        if variant.fields.style != darling::ast::Style::Tuple || variant.fields.fields.len() != 1 {
            return Err(syn::Error::new(
                variant.span(),
                "InlineFragments derive requires enum variants to have one unnamed field",
            ));
        }
        let field = variant.fields.fields.first().unwrap();
        result.push(Fragment {
            rust_variant_name: variant.ident.clone(),
            inner_type: field.ty.clone(),
        });
    }
    Ok(result)
}

fn check_fallback(
    variants: &[SpannedValue<InlineFragmentsDeriveVariant>],
    target_type: &InlineFragmentType,
) -> Result<Option<(syn::Ident, Option<syn::Type>)>, Errors> {
    let fallbacks = variants.iter().filter(|v| *v.fallback).collect::<Vec<_>>();

    if fallbacks.is_empty() {
        return Ok(None);
    }

    if fallbacks.len() > 1 {
        let mut errors = Errors::default();
        for fallback in &fallbacks[1..] {
            errors.push(syn::Error::new(
                fallback.span(),
                "InlineFragments can't have more than one fallback",
            ))
        }

        return Err(errors);
    }

    let fallback = fallbacks[0];
    match target_type {
        InlineFragmentType::Interface(_) => match fallback.fields.style {
            darling::ast::Style::Struct => Err(syn::Error::new(
                fallback.span(),
                "InlineFragment fallbacks don't currently support struct variants",
            )
            .into()),
            darling::ast::Style::Tuple => {
                if fallback.fields.len() != 1 {
                    return Err(syn::Error::new(
                        fallback.span(),
                        "InlineFragments require variants to have one unnamed field",
                    )
                    .into());
                }
                Ok(Some((
                    fallback.ident.clone(),
                    Some(fallback.fields.fields[0].ty.clone()),
                )))
            }
            darling::ast::Style::Unit => Ok(Some((fallback.ident.clone(), None))),
        },
        InlineFragmentType::Union(_) => {
            if fallback.fields.style != darling::ast::Style::Unit {
                return Err(syn::Error::new(
                    fallback.span(),
                    "The fallback for a union type must be a unit variant",
                )
                .into());
            }
            Ok(Some((fallback.ident.clone(), None)))
        }
    }
}

struct QueryFragmentImpl<'a> {
    target_enum: syn::Ident,
    type_lock: syn::Path,
    argument_struct: syn::Type,
    fragments: &'a [Fragment],
    graphql_type_name: String,
    fallback: Option<(syn::Ident, Option<syn::Type>)>,
}

impl quote::ToTokens for QueryFragmentImpl<'_> {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        use quote::{quote, TokenStreamExt};

        let target_struct = &self.target_enum;
        let type_lock = &self.type_lock;
        let arguments = &self.argument_struct;
        let inner_types: Vec<_> = self
            .fragments
            .iter()
            .map(|fragment| &fragment.inner_type)
            .collect();
        let graphql_type = proc_macro2::Literal::string(&self.graphql_type_name);
        let fallback_selection = match &self.fallback {
            Some((_, Some(fallback_fragment))) => quote! {
                <#fallback_fragment>::query(builder);
            },
            _ => quote! {},
        };

        tokens.append_all(quote! {
            #[automatically_derived]
            impl<'de> ::cynic::QueryFragment<'de> for #target_struct {
                type SchemaType = #type_lock;
                type Variables = #arguments;

                const TYPE: Option<&'static str> = Some(#graphql_type);

                fn query(mut builder: ::cynic::queries::SelectionBuilder<Self::SchemaType, Self::Variables>) {
                    #(
                        let fragment_builder = builder.inline_fragment();
                        let mut fragment_builder = fragment_builder.on::<<#inner_types as ::cynic::QueryFragment>::SchemaType>();
                        <#inner_types as ::cynic::QueryFragment>::query(
                            fragment_builder.select_children()
                        );
                    )*

                    #fallback_selection
                }
            }
        })
    }
}

fn find_union_or_interface_type<'a>(
    name: &str,
    schema: &'a schema::Document,
) -> Option<InlineFragmentType<'a>> {
    for definition in &schema.definitions {
        use graphql_parser::schema::{Definition, TypeDefinition};
        match definition {
            Definition::TypeDefinition(TypeDefinition::Union(union)) => {
                if union.name == name {
                    return Some(InlineFragmentType::Union(union));
                }
            }
            Definition::TypeDefinition(TypeDefinition::Interface(interface)) => {
                if interface.name == name {
                    return Some(InlineFragmentType::Interface(interface));
                }
            }
            _ => {}
        }
    }

    None
}

enum InlineFragmentType<'a> {
    Union(&'a schema::UnionType),
    Interface(&'a schema::InterfaceType),
}

impl<'a> InlineFragmentType<'a> {
    pub fn name(&self) -> &str {
        match self {
            InlineFragmentType::Interface(iface) => &iface.name,
            InlineFragmentType::Union(union) => &union.name,
        }
    }
}
