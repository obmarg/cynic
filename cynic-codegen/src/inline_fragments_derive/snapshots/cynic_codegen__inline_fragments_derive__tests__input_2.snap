---
source: cynic-codegen/src/inline_fragments_derive/tests.rs
assertion_line: 56
expression: "format_code(format!(\"{}\", tokens))"

---
#[automatically_derived]
impl<'de> ::cynic::serde::Deserialize<'de> for PostOrAuthor {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: ::cynic::serde::Deserializer<'de>,
    {
        deserializer.deserialize_map(::cynic::__private::InlineFragmentVisitor::<Self>::new())
    }
}
#[automatically_derived]
impl<'de> ::cynic::InlineFragments<'de> for PostOrAuthor {
    fn deserialize_variant<D>(typename: &str, deserializer: D) -> Result<Self, D::Error>
    where
        D: ::cynic::serde::Deserializer<'de>,
    {
        if Some(typename) == <Post as ::cynic::QueryFragment<'de>>::TYPE {
            return <Post as ::cynic::serde::Deserialize<'de>>::deserialize(deserializer)
                .map(PostOrAuthor::Post);
        }
        if Some(typename) == <Author as ::cynic::QueryFragment<'de>>::TYPE {
            return <Author as ::cynic::serde::Deserialize<'de>>::deserialize(deserializer)
                .map(PostOrAuthor::Author);
        }
        Ok(PostOrAuthor::Other)
    }
}
#[automatically_derived]
impl<'de> ::cynic::QueryFragment<'de> for PostOrAuthor {
    type SchemaType = schema::PostOrAuthor;
    type Variables = ();
    const TYPE: Option<&'static str> = Some("PostOrAuthor");
    fn query(mut builder: ::cynic::queries::SelectionBuilder<Self::SchemaType, Self::Variables>) {
        let fragment_builder = builder.inline_fragment();
        let mut fragment_builder =
            fragment_builder.on::<<Post as ::cynic::QueryFragment>::SchemaType>();
        <Post as ::cynic::QueryFragment>::query(fragment_builder.select_children());
        let fragment_builder = builder.inline_fragment();
        let mut fragment_builder =
            fragment_builder.on::<<Author as ::cynic::QueryFragment>::SchemaType>();
        <Author as ::cynic::QueryFragment>::query(fragment_builder.select_children());
    }
}

